[subclasses] Fix nested subclasses flattened tensors ordering

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132096

get_plain_tensors() should result in DFS of leaves.
The error was that plain tensors (leaves) on the same level were returned before subclasses plained tensors even if subclasses are before in "flatten" list.

Original issue from AO: https://github.com/pytorch/ao/issues/515

Test:TBD, need to make asymetric subclass with dense tensors and subclasses

IvanKobzarev committed:
[subclasses] Fix nested subclasses flattened tensors ordering

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -156,13 +156,14 @@ def get_plain_tensors(subclass: Tensor) -> List[Tensor]:
     todo = [subclass]
     while todo:
         curr = todo.pop()
+        if not is_traceable_wrapper_subclass(curr):
```
cc @tugsbayasgalan - the problem was that we are desugaring nested subclasses in breadth-first order, which means that inner tensors corresponding to the same top-level subclass will not show up next to each other in the final desugared List[Tensor]. But `wrap_tensor_subclasses()` in AOTAutograd requires this to be true.

bdhirsh reviewed (approved):
landing to unblock AO sounds ok, although a followup test would be good to have

IvanKobzarev committed:
Update on "[subclasses] Fix nested subclasses flattened tensors ordering"


get_plain_tensors() should result in DFS of leaves.
The error was that plain tensors (leaves) on the same level were returned before subclasses plained tensors even if subclasses are before in "flatten" list.

Original issue from AO: https://github.com/pytorch/ao/issues/515

Test:TBD, need to make asymetric subclass with dense tensors and subclasses

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
